rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,444 @@
+// Copyright (C) 2019-2020 Siodb GmbH. All rights reserved.
+// Use of this source code is governed by a license that can be found
+// in the LICENSE file.
+
+package main
+
+import (
+	"SiodbIomgrProtocol"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang/protobuf/proto"
+)
+
+var (
+	IOMgrChunkMaxSize    uint32 = 65536
+	MessageLengthMaxSize uint32 = 1 * 1024 * 1024
+	IOMgrReadDeadline           = 5 * time.Second
+)
+
+var (
+	DATABASEENGINERESPONSE    uint32 = 4
+	DATABASEENGINERESTREQUEST uint32 = 13
+)
+
+type IOMgrConnection struct {
+	*TrackedNetConn
+	pool *IOMgrConnPool
+}
+
+func (IOMgrConn *IOMgrConnection) cleanupTCPConn() (err error) {
+
+	var IOMgrChunkSize uint32
+	var counter int16
+
+	for {
+		// Get Current Row Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint32(); err != nil {
+			return fmt.Errorf("cleanupTCPConn: unable to read chunk size")
+		}
+		if IOMgrChunkSize == 0 {
+			siodbLoggerPool.Debug("cleanupTCPConn: %d chunks dropped.", counter)
+			return err
+		}
+		siodbLoggerPool.Debug("cleanupTCPConn: Chunk size to drop: %d.", IOMgrChunkSize)
+		buff := make([]byte, IOMgrChunkSize)
+		IOMgrConn.SetReadDeadline(time.Now().Add(IOMgrReadDeadline))
+		_, err = io.ReadFull(IOMgrConn, buff)
+		if err != nil {
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				IOMgrConn.Close()
+				IOMgrConn.TrackedNetConn.Conn = nil
+				return fmt.Errorf("cleanupTCPConn | Protocol error (read timeout): %v", err)
+			} else {
+				return fmt.Errorf("cleanupTCPConn | Protocol error: %v", err)
+			}
+		}
+
+		counter++
+	}
+}
+
+func (IOMgrConn *IOMgrConnection) writeJSONPayload(requestID uint64, c *gin.Context) (err error) {
+
+	// Read Payload up to max json payload size
+	var bytesRead int = 0
+	var bytesReadTotal int = 0
+	var bytesWritten int = 0
+	var bytesWrittenTotal int = 0
+	buffer := make([]byte, restServerConfig.RequestPayloadBufferSize)
+
+	siodbLoggerPool.Debug("writeJSONPayload | IOMgrConn.pool.maxJsonPayloadSize: %v, restServerConfig.RequestPayloadBufferSize: %v",
+		IOMgrConn.pool.maxJsonPayloadSize, restServerConfig.RequestPayloadBufferSize)
+
+	for {
+		if uint64(bytesReadTotal) > IOMgrConn.pool.maxJsonPayloadSize {
+			break
```

Comment:
@cxxman if I return error straigh after the condition, the IOMgr will expecting me to read the response. Thus should I read the response message and raise an error or close the connection?

Reply:
close connection

Reply:
Ok, implemented. I found that when I close the failed connection while IOMgr was reading the payload, IOMgr crashed:

```bash
2020-09-30 12:22:53.376368 info 2249628 2249634 IPv4-RESTConnectionManager: Accepted new TCP connection from 2.0.182.46.
2020-09-30 12:22:53.376530 debug 2249628 2249638 IOManagerConnectionHandler-1[563]: Waiting for request...
2020-09-30 12:22:53.376581 info 2249628 2249638 Instance: User 'ROOT' authenticated via token.
2020-09-30 12:22:53.376653 debug 2249628 2249638 IOManagerConnectionHandler-1[563]: Creating DBEngineRestRequest
2020-09-30 12:22:53.376682 debug 2249628 2249638 DBEngineRestRequestFactory: Reading JSON payload
terminate called after throwing an instance of 'siodb::net::ConnectionError'
  what():  ProtobufInputStream: Connection closed
2020-09-30 12:22:54.547148 warning 2249554 2249555 ProcessMonitor: IO Manager (PID 2249628) has unexpectedly exited with status 0
2020-09-30 12:22:54.547224 error 2249554 2249555 ProcessMonitor: IO Manager has been restarted too many times in a period of 8 seconds. This may indicate a persistent issue. Giving up on restarting IO Manager.
2020-09-30 12:22:54.547290 info 2249554 2249555 ProcessMonitor: IO Manager monitor thread stopped.
2020-09-30 12:22:54.547644 info 2249554 2249554 Database instance is shutting down due to signal #2 (Interrupt).
2020-09-30 12:22:54.548211 info 2249554 2249567 IPv4-SiodbConnectionManager: TCP connection listener thread is exiting because database is shutting down.
2020-09-30 12:22:54.548666 info 2249554 2249565 UNIX-SiodbConnectionManager: UNIX connection listener thread is exiting because database is shutting down.
2020-09-30 12:22:54.548978 info 2249554 2249554 ProcessMonitor: Shutting down.
2020-09-30 12:22:54.549157 info 2249554 2249554 ProcessMonitor: Stopping IO Manager monitor thread
2020-09-30 12:22:54.549313 info 2249554 2249554 ProcessMonitor: IO Manager monitor thread stopped.
2020-09-30 12:22:54.549456 info 2249554 2249554 Logging stopped.
```

Reply:
ok, this will require fix in the iomgr, let's test this case after fix.